src/coreclr/jit/morph.cpp

```diff
@@ -12531,11 +12531,6 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             }
 #endif
 #endif // !TARGET_64BIT
-
-            if (op2->gtOper == GT_CAST && op2->AsOp()->gtOp1->IsCnsIntOrI())
```

Comment:
Since we fold it in importer, can we add an assert that we don't see that particular const cast here and if it does, we would know that the caller has to fix it?

```c++
assert(op2->gtOper != GT_CAST || !op2->AsOp()->gtOp1->IsCnsIntOrI());
```

Reply:
Would it be OK for me to add the assert in the next PR (that will eliminate the implicit upcasts done for 64 bit operands in importer)? Since this change is quite old, I'd have to rebase to re-run the replays and make sure we're not seeing cases where it fires.

Note: missing this on 64 bit is not a problem because we will fold it anyway in post-order. The reason there is this non-trivial amount of code in the pre-order traversal that tries to look for folding opportunities is because it is the "last chance" to see them before morphing into helpers.

Reply:
sounds good.